ddev/ddev
# fix: Update to Traefik v3, make Let's Encrypt work with Traefik, fixes #6183, for #5940, fixes #4632

## The Issue

* #6183 
* #4632 
* #5940 (on the way to deprecation of nginx-proxy router)

## How This PR Solves The Issue

* Update to v3 of traefik
* Make Let's Encrypt work with Traefik

## TODO

- [x] When `use_letsencrypt` we have to delete the default certs
- [x] Generated .gitignore should detect traefik configuration that has had #ddev-generated removed
- [x] Consider looking at changed Traefik v3 config directives and updating them.
- [x] Review generated project-level configs for accuracy. Can the routers or services be consolidated?
- [x] Review how push to volume happens. Do we need to empty some things before push?
- [x] We should be able to have just the one http service, not http + https
- [x] Verify DNS challenge docs
- [x] Add suggested summaries in each section of the traefik doc
- [x] static_config.yaml either must be off-limits for custom configuration, or `entryPoints` has to be managed elsewhere and merged into it.
- [x] Since static_config.yaml can't really safely be edited, certificateResolvers should be in there by default for Let's Encrypt
- [x] Since static_config.yaml can't really be safely edited, changing it for the Let's Encrypt DNS challenge isn't going to be a win I don't think. Or perhaps it can just have a different name and be included by default.
- [x] docs need to say that we're using Traefik v2 syntax
- [x] docs: Consider mentioning that additional files can be added in .ddev/traefik/config; filenames must be unique (base on project name) and stanza names must be unique (base on project name).
- [x] Support Traefik plugins in static_config.yaml

## Followups
- [ ] Try to merge static_config instead of appending. Traefik is unforgiving about this with static config, even though it's not bad with dynamic config.
- [ ] After that is solved, add information back to docs about using the Let's Encrypt staging server

## Manual Testing Instructions

* Start by pulling latest traefik image: `docker pull ddev/ddev-traefik-router:20240616_traefik_3`
* General Usage
  * Try basic use and see if things work
  * Add add-ons and check that all the ports referenced in them work correctly, including http and https ports. For example try `ddev get ddev/ddev-solr` and see that solr UI works right. 

* Inspect the created configuration files
  * When doing an add-on or whatever above, review the project-generated files in .ddev/traefik/config to see if they make sense.
  * Edit a file to add new behaviors like redirect or adding a header with middleware. `ddev restart` and verify it works (have to remove #ddev-generated). Don't forget to remove the file then so it will be recreated.
* Try out casual hosting with this PR
* Try out dnsChallenge docs with traefik (this worked for me, but may not be useful to pursue. It does require extra capabilities and permissions on the DNS setup)

Traefik configuration docs at https://ddev--6317.org.readthedocs.build/en/6317/users/extend/traefik-router/
Casual hosting docs at https://ddev--6317.org.readthedocs.build/en/6317/users/topics/hosting/

## Automated Testing Overview

There isn't currently testing for Casual hosting
I think the existing tests will catch problems with traefik v3. I imagine there will be some

## Release/Deployment Notes

* Customized Traefik rules will likely be affected by this PR and may have to be adjusted.
* Traefik is updated to v3 but still uses v2 configuration
